NumLib/Fem/Deformation/LinearBMatrix.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_FEM_DEFORMATION_LINEARBMATRIX_H_
+#define NUMLIB_FEM_DEFORMATION_LINEARBMATRIX_H_
+
+#include <cmath>
+
+namespace NumLib
+{
+namespace LinearBMatrix
+{
+/// Fills a B-matrix based on given shape function dN/dx values.
+template <int DisplacementDim, int NPOINTS, typename DNDX_Type,
+          typename BMatrixType>
+void computeBMatrix(DNDX_Type const& dNdx, BMatrixType& b_matrix)
+{
+    static_assert(0 < DisplacementDim && DisplacementDim <= 3,
+                  "LinearBMatrix::computeBMatrix: DisplacementDim must be in "
+                  "range [1,3].");
+
+    b_matrix.setZero();
+
+    switch (DisplacementDim)
+    {
+        case 3:
+        {
+            for (int i = 0; i < NPOINTS; ++i)
+            {
+                b_matrix(2, 2 * NPOINTS + i) = dNdx(2, i);
+                b_matrix(4, NPOINTS + i) = dNdx(2, i) / std::sqrt(2);
+                b_matrix(4, 2 * NPOINTS + i) = dNdx(1, i) / std::sqrt(2);
+                b_matrix(5, i) = dNdx(2, i) / std::sqrt(2);
+                b_matrix(5, 2 * NPOINTS + i) = dNdx(0, i) / std::sqrt(2);
+            }
+        }
+        case 2:
+        {
+            for (int i = 0; i < NPOINTS; ++i)
+            {
+                b_matrix(1, NPOINTS + i) = dNdx(1, i);
+                b_matrix(3, i) = dNdx(1, i) / std::sqrt(2);
+                b_matrix(3, NPOINTS + i) = dNdx(0, i) / std::sqrt(2);
+            }
+        }
+        // no break for fallthrough.
+        case 1:
+        {
+            for (int i = 0; i < NPOINTS; ++i)
+                b_matrix(0, i) = dNdx(0, i);
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+/// Compute strain from given displacment and B-matrix.
+template <typename NodalDisplacementVectorType, typename BMatrixType,
+          typename StrainVectorType>
+void computeStrain(NodalDisplacementVectorType const& local_displacement,
```

Comment:
Using this function it would be very easy to test the correctness of the BMatrix implementation, I guess.